overseer/src/lib.rs

```diff
@@ -0,0 +1,623 @@
+	s: S,
+
+	/// Here we keep handles to spawned subsystems be notified when they terminate.
+	running_subsystems: FuturesUnordered<RemoteHandle<()>>,
+}
+
+impl<M, S, I> Overseer<M, S, I>
+where
+	M: Debug + Clone,
+	S: Spawn,
+	I: Eq + Copy + Debug + std::hash::Hash,
+{
+	/// Create a new intance of the `Overseer` with some initial set of `Subsystems.
+	///
+	/// The `Subsystems` submitted to this call will act as a level 1 in the "process tree":
+	///
+	///
+	/// ```text
+	///                  +------------------------------------+
+	///                  |            Overseer                |
+	///                  +------------------------------------+
+	///                    /            |             |      \
+	///      ................. subsystems[..] ..............................
+	///      . +-----------+    +-----------+   +----------+   +---------+ .
+	///      . |           |    |           |   |          |   |         | .
+	///      . +-----------+    +-----------+   +----------+   +---------+ .
+	///      ...............................................................
+	///                              |
+	///                        probably `spawn`
+	///                        something else
+	///                              |
+	///                              V
+	///                         +-----------+
+	///                         |           |
+	///                         +-----------+
+	///
+	/// ```
+	pub fn new<T: IntoIterator<Item = (I, Box<dyn Subsystem<M, I> + Send>)>>(subsystems: T, s: S) -> Self {
+		let mut this = Self {
+			subsystems: HashMap::new(),
+			id_to_children: HashMap::new(),
+			s,
+			running_subsystems: FuturesUnordered::new(),
+		};
+
+		for s in subsystems.into_iter() {
+			let _ = this.spawn(s);
+		}
+
+		this
+	}
+
+	/// Run the `Overseer`.
+	// TODO: we have to
+	//   * Give out to the user some handler to communicate with the `Overseer`
+	//     to tell it to do things such as `Start` `Stop` or `Spawn`
+	//   * Actually implement stopping of the `Overseer`, atm it's unstoppable.
+	pub async fn run(mut self) {
+		loop {
+			// Upon iteration of the loop we will be collecting all the messages
+			// that need dispatching (if any).
+			let mut msgs = Vec::default();
+
+			for (id, s) in self.subsystems.iter_mut() {
+				if let Some(s) = &mut s.instance {
+					while let Poll::Ready(Some(msg)) = poll!(&mut s.rx.next()) {
+						log::info!("Received message from subsystem {:?}", msg);
+						msgs.push((*id, msg));
+					}
+				}
+			}
+
+			// Do the message dispatching be it broadcasting or direct messages.
+			//
+			// TODO: this desperately need refactoring.
+			for msg in msgs.into_iter() {
+				match msg.1 {
+					OverseerMessage::SubsystemMessage{ to, msg: m } => {
+						match to {
+							Some(to) => {
+								if let Some(subsystem) = self.subsystems.get_mut(&to) {
+									if let Some(ref mut i) = subsystem.instance {
+										let _ = i.tx.send(m).await;
```

Comment:
Do I understand correctly that if a singe subsystem is pending (e.g. waiting for some timeout) and thus does not empty its receive channel, all other subsystems are blocked as well as it will block at this `await`?

I am not sure whether this would actually be a problem. I guess this would depend on the design of the individual subsystems.

Reply:
yes